iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp

```diff
@@ -292,12 +110,27 @@ struct ConvertToFlowTensorOpsPass
     FuncOp funcOp = getOperation();
     MLIRContext *context = funcOp->getContext();
     context->allowUnregisteredDialects(true);
+
+    // Run partial conversion from tensor -> flow.tensor ops.
```

Comment:
Seems like this refactoring is changing some preconditions that the `linalg_ext.sort` lowering relied on (for dynamic shapes?), but I'm not sure yet what the specific issue is. This change should only be moving patterns around, it does not change the details within any patterns.

Before:
* ConvertUpstreamToIREE for `tensor.cast` and `tensor.from_elements`
* ConvertToFlowTensorOps for `tensor.insert_slice` and `tensor.extract_slice`

After:
* ConvertToFlowTensorOps for `tensor.cast`, `tensor.from_elements`, `tensor.insert_slice`, and `tensor.extract_slice`

Previously `memref::createResolveShapedTypeResultDimsPass` was removing `tensor.cast` ops, but now they're already converted to `flow.tensor.reshape`

@hanhanW or @MaheshRavishankar any ideas? I can make this refactor more of an NFC, but organizing all tensor->flow patterns together seemed like a good idea to me. Detensoring will also want the `tensor.from_elements` patterns, so I at least want to put those in a spot other than `ConvertUpstreamToIREE`

Reply:
I had a chat with Mahesh offline.

* The pass should not be converting `tensor.cast` ops into `flow.tensor.reshape` ops within dispatch regions - flow ops should only exist outside of dispatch regions.
* I could split the patterns to get this change back to an NFC (make `createConvertToFlowTensorOpsPass(true)` and `createConvertToFlowTensorOpsPass(false)` match their prior behavior, expose patterns that were in `ConvertUpstreamToIREE` through a different mechanism)
* Writing lit tests for the `// Elementwise, fusion, tiling and distribution.` block of passes would be tricky. Lots of optional optimizations and specific sequencing. Maybe we could verify no flow ops within dispatch regions though? (I'd like some more specific error message for my incorrect pass compared to this e2e sort.mlir test failing and needing to dig through why)

Reply:
(no [magic booleans](https://blog.codinghorror.com/avoiding-booleans/) - have a different pass name or an enum or something meaningful)

Reply:
Oh, the booleans came from https://github.com/google/iree/pull/5614 :P 

I'd be doing something more like this for the second bullet point
```diff
void populateTensorToFlowPatterns(MLIRContext *context,
                                  OwningRewritePatternList &patterns,
                                  TypeConverter &typeConverter);

+void populateDispatchTensorToFlowPatterns(MLIRContext *context,
+                                          OwningRewritePatternList &patterns,
+                                          TypeConverter &typeConverter);
```

Going to first see if only operating outside of dispatch regions (the first bullet) is sufficient.

Reply:
Nice, I think that change worked, no need to split the patterns:

```c++
  static bool shouldBeConverted(tensor::CastOp op) {
    if (op->getParentOfType<Flow::DispatchWorkgroupsOp>()) {
      return false;
    }

    return true;
  }
```

I'll get this ready for review next week. I should add some unit tests for those `shouldBeConverted` checks. It looks like _some_ flow ops can exist within dispatch regions, but maybe a selective verifier could still help spot issues like this earlier?